### DISCUSSION TITLE
Polled versus Interrupt driven Touchpanel

### DISCUSSION THREAD

**Discussion by memeticae** [Bodmer/TFT_eSPI#2272] [General]
Hello,

I've been using this library for some time now, to develop an application whilst using an esp32-D0WD.
And it works with an MHS4.0-rev.B 320x480 display (ST7796), with an xpt2046 touch controller.

With vscode, platformio, the espressif32 platform, and arduino framework(), I managed to put some pictures on the screen.

However I would like to see an interrupt driven routine that handles touch input. 
(Because I'm sleeping al the time.)

This particular screen has an #TP_INT signal on (its) pin 11.

How can I write a procedure to get the "real" values of the "tft.touchvalues" using the  tft.gettouch() function?
Without having to implement it myself?

**Comment by Bodmer** [comment]
The library does not support interrupts. Options:
1. Connect the interrupt, use it to wake the processor then use polling to handle touch actions.
2. Use another touch library that supports interrupts. In this case use a different SPI bus for the touch interface to avoid conflicts with screen access.

**Comment by Lebo77** [comment]
I just did option 1. 

Hooked the hardware interrupt on the display to pull down a digital input pin on the microcontroller, programmed the microcontroller to trigger the Interrupt Service Routine (ISR)  when it saw a falling edge. The ISR just set a semaphore and returned. (You want to keep ISRs as short as possible.)

```
SemaphoreHandle_t touchSemaphore = NULL; // handle for the touch semaphore

void IRAM_ATTR touchISR() {
  // Unblock the task by releasing the semaphore.
  xSemaphoreGiveFromISR( touchSemaphore, nullptr );
}

void main function () {

  pinMode(TOUCH_IRQ_PIN, INPUT_PULLUP);
  attachInterrupt(TOUCH_IRQ_PIN, touchISR, FALLING);
...
for (;;) {
    if ( xSemaphoreTake( touchSemaphore, 200 / portTICK_PERIOD_MS ) == pdTRUE ) // bail out to reset the WDT every 1/5th sec. and for press and hold tracking.
    { ... }
else {}
}
```

The Semaphore in the for(;;) loop says (wit for this semaphore to be available and when it is run the stuff in my conditional block. If 200 milliseconds pass and nothing has triggered me, run the stuff in the "else" block (reset watchdog timers and such).

**Comment by Bwanna** [comment]
Lebo77,
I'm also trying to incorporate Touch IRQ when using TFT_eSPI and want to try this approach.  Question if you don't mind?  What would I need to do for your code below to run the normal touchscreen response as if it had been triggered by the polling approach?   How do I insert this to work in the normal TFT_eSPI examples?

Thanks for the assistance.



> I just did option 1.
> 
> Hooked the hardware interrupt on the display to pull down a digital input pin on the microcontroller, programmed the microcontroller to trigger the Interrupt Service Routine (ISR) when it saw a falling edge. The ISR just set a semaphore and returned. (You want to keep ISRs as short as possible.)
> 
> ```
> SemaphoreHandle_t touchSemaphore = NULL; // handle for the touch semaphore
> 
> void IRAM_ATTR touchISR() {
>   // Unblock the task by releasing the semaphore.
>   xSemaphoreGiveFromISR( touchSemaphore, nullptr );
> }
> 
> void main function () {
> 
>   pinMode(TOUCH_IRQ_PIN, INPUT_PULLUP);
>   attachInterrupt(TOUCH_IRQ_PIN, touchISR, FALLING);
> ...
> for (;;) {
>     if ( xSemaphoreTake( touchSemaphore, 200 / portTICK_PERIOD_MS ) == pdTRUE ) // bail out to reset the WDT every 1/5th sec. and for press and hold tracking.
>     { ... }
> else {}
> }
> ```
> 
> The Semaphore in the for(;;) loop says (wit for this semaphore to be available and when it is run the stuff in my conditional block. If 200 milliseconds pass and nothing has triggered me, run the stuff in the "else" block (reset watchdog timers and such).
